File: android/src/main/java/expo/modules/checkoutmodule/Appearance.kt
package expo.modules.checkoutmodule

import com.checkout.components.interfaces.uicustomisation.BorderRadius
import com.checkout.components.interfaces.uicustomisation.designtoken.ColorTokens
import com.checkout.components.interfaces.uicustomisation.designtoken.DesignTokens
import com.checkout.components.interfaces.uicustomisation.font.Font
import com.checkout.components.interfaces.uicustomisation.font.FontName

/**
 * Checkout Flow DesignTokens from JS.
 * Shape: { colorTokens: { action?, background?, ... }, borderRadius?: { all } | { radius, corners }, borderFormRadius?, fonts?: { button?, input?, ... } }
 * Also accepts legacy "colors" in place of "colorTokens".
 * Converts to SDK [DesignTokens] for [CheckoutComponentConfiguration.appearance].
 */
data class CheckoutAppearanceOptions(
    val colorTokens: Map<String, Long> = emptyMap(),
    val borderRadiusAll: Int? = null,
    val borderRadiusRadius: Int = 4,
    val borderRadiusCorners: List<String> = emptyList(),
    val borderFormRadiusAll: Int? = null,
    val borderFormRadiusRadius: Int = 4,
    val borderFormRadiusCorners: List<String> = emptyList(),
    val fonts: Map<String, FontDescriptor> = emptyMap(),
) {
    data class FontDescriptor(
        val fontFamily: String? = null,
        val fontSize: Double = 17.0,
        val fontWeight: String? = null,
        val fontStyle: String? = null,
        val letterSpacing: Double? = null,
        val lineHeight: Double? = null,
    )

    companion object {
        const val DEFAULT_ACTION = 0xFF186AFFL
        const val DEFAULT_BACKGROUND = 0xFFFFFFFFL
        const val DEFAULT_BORDER = 0xFFDDDDDDL
        const val DEFAULT_DISABLED = 0xFFB0B0B0L
        const val DEFAULT_ERROR = 0xFFAD283EL
        const val DEFAULT_FORM_BACKGROUND = 0xFFFFFFFFL
        const val DEFAULT_FORM_BORDER = 0xFF949494L
        const val DEFAULT_INVERSE = 0xFFFFFFFFL
        const val DEFAULT_OUTLINE = 0xFF91B2EEL
        const val DEFAULT_PRIMARY = 0xFF000000L
        const val DEFAULT_SECONDARY = 0xFF727272L
        const val DEFAULT_SUCCESS = 0xFF00856AL
        const val DEFAULT_SCROLLED_CONTAINER = 0xFFE8E6E6L

        /** Parses hex string "#RRGGBB" or "#AARRGGBB" to 0xAARRGGBB Long. */
        private fun parseHex(hex: String?): Long? {
            if (hex.isNullOrBlank()) return null
            val s = hex.trim().removePrefix("#")
            if (s.length != 6 && s.length != 8) return null
            return s.toLongOrNull(16)?.let { if (s.length == 6) 0xFF000000L or it else it }
        }

        private fun parseRadiusDict(map: Map<*, *>?): Triple<Int?, Int, List<String>> {
            if (map == null) return Triple(null, 4, emptyList())
            val all = (map["all"] as? Number)?.toInt()
            if (all != null) return Triple(all, all, emptyList())
            val radius = (map["radius"] as? Number)?.toInt() ?: 4
            val corners = (map["corners"] as? List<*>)?.mapNotNull { it as? String } ?: emptyList()
            return Triple(null, radius, corners)
        }

        private fun parseFontDescriptor(map: Map<*, *>?): FontDescriptor? {
            if (map == null) return null
            return FontDescriptor(
                fontFamily = map["fontFamily"] as? String,
                fontSize = (map["fontSize"] as? Number)?.toDouble() ?: 17.0,
                fontWeight = map["fontWeight"] as? String,
                fontStyle = map["fontStyle"] as? String,
                letterSpacing = (map["letterSpacing"] as? Number)?.toDouble(),
                lineHeight = (map["lineHeight"] as? Number)?.toDouble(),
            )
        }

        /**
         * Parses JS DesignTokens (Appearance) into [CheckoutAppearanceOptions].
         * Returns null if neither colorTokens nor colors present.
         */
        fun fromMap(options: Map<String, Any?>?): CheckoutAppearanceOptions? {
            val colorTokensDict = (options?.get("colorTokens") as? Map<*, *>) ?: (options?.get("colors") as? Map<*, *>) ?: return null
            val colorStrings = colorTokensDict.mapNotNull { (k, v) ->
                (k as? String)?.let { key -> (v as? String)?.let { key to it } }
            }.toMap()

            val colorTokens = colorStrings.mapValues { (_, hex) ->
                parseHex(hex) ?: DEFAULT_ACTION
            }.toMutableMap()
            if (colorTokens["action"] == null) colorTokens["action"] = DEFAULT_ACTION
            if (colorTokens["background"] == null) colorTokens["background"] = DEFAULT_BACKGROUND
            if (colorTokens["border"] == null) colorTokens["border"] = DEFAULT_BORDER
            if (colorTokens["disabled"] == null) colorTokens["disabled"] = DEFAULT_DISABLED
            if (colorTokens["error"] == null) colorTokens["error"] = DEFAULT_ERROR
            if (colorTokens["formBackground"] == null) colorTokens["formBackground"] = DEFAULT_FORM_BACKGROUND
            if (colorTokens["formBorder"] == null) colorTokens["formBorder"] = DEFAULT_FORM_BORDER
            if (colorTokens["inverse"] == null) colorTokens["inverse"] = DEFAULT_INVERSE
            if (colorTokens["outline"] == null) colorTokens["outline"] = DEFAULT_OUTLINE
            if (colorTokens["primary"] == null) colorTokens["primary"] = DEFAULT_PRIMARY
            if (colorTokens["secondary"] == null) colorTokens["secondary"] = DEFAULT_SECONDARY
            if (colorTokens["success"] == null) colorTokens["success"] = DEFAULT_SUCCESS

            val (borderRadiusAll, borderRadiusRadius, borderRadiusCorners) = parseRadiusDict(options?.get("borderRadius") as? Map<*, *>)
            val (borderFormRadiusAll, borderFormRadiusRadius, borderFormRadiusCorners) = parseRadiusDict(
                options?.get("borderFormRadius") as? Map<*, *> ?: options?.get("borderRadius") as? Map<*, *>
            )

            val fontsDict = options?.get("fonts") as? Map<*, *> ?: emptyMap<String, Any?>()
            val fonts = listOf("button", "footnote", "input", "label", "subheading").mapNotNull { key ->
                parseFontDescriptor(fontsDict[key] as? Map<*, *>)?.let { key to it }
            }.toMap()

            return CheckoutAppearanceOptions(
                colorTokens = colorTokens,
                borderRadiusAll = borderRadiusAll,
                borderRadiusRadius = borderRadiusRadius,
                borderRadiusCorners = borderRadiusCorners,
                borderFormRadiusAll = borderFormRadiusAll,
                borderFormRadiusRadius = borderFormRadiusRadius,
                borderFormRadiusCorners = borderFormRadiusCorners,
                fonts = fonts,
            )
        }
    }

    /**
     * Converts to SDK [DesignTokens] for [CheckoutComponentConfiguration.appearance].
     * SDK types live in com.checkout.components.interfaces.uicustomisation(.designtoken).
     */
    fun toSdkDesignTokens(): DesignTokens {
        val c = colorTokens
        val colorTokensSdk = ColorTokens(
            (c["action"] ?: DEFAULT_ACTION),
            (c["background"] ?: DEFAULT_BACKGROUND),
            (c["border"] ?: DEFAULT_BORDER),
            (c["disabled"] ?: DEFAULT_DISABLED),
            (c["error"] ?: DEFAULT_ERROR),
            (c["formBackground"] ?: DEFAULT_FORM_BACKGROUND),
            (c["formBorder"] ?: DEFAULT_FORM_BORDER),
            (c["inverse"] ?: DEFAULT_INVERSE),
            (c["outline"] ?: DEFAULT_OUTLINE),
            (c["primary"] ?: DEFAULT_PRIMARY),
            (c["secondary"] ?: DEFAULT_SECONDARY),
            (c["success"] ?: DEFAULT_SUCCESS),
            (c["scrolledContainer"] ?: DEFAULT_SCROLLED_CONTAINER),
        )
        val radiusButton = borderRadiusAll ?: borderRadiusRadius
        val radiusForm = borderFormRadiusAll ?: borderFormRadiusRadius
        val borderRadiusSdk = BorderRadius(radiusButton)
        val borderFormRadiusSdk = BorderRadius(radiusForm)
        return DesignTokens(
            colorTokens = colorTokensSdk,
            emptyMap<FontName, Font>(), // custom fonts not mapped from JS yet
            borderFormRadiusSdk,
            borderRadiusSdk,
        )
    }
}

File: ios/Appearance.swift
import SwiftUI
import CheckoutComponentsSDK

extension Color {
  init(hex: String) {
    let hex = hex.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
    var int: UInt64 = 0
    Scanner(string: hex).scanHexInt64(&int)
    let a, r, g, b: UInt64
    switch hex.count {
    case 3: (a, r, g, b) = (255, (int >> 8) * 17, (int >> 4 & 0xF) * 17, (int & 0xF) * 17)
    case 6: (a, r, g, b) = (255, int >> 16, int >> 8 & 0xFF, int & 0xFF)
    case 8: (a, r, g, b) = (int >> 24, int >> 16 & 0xFF, int >> 8 & 0xFF, int & 0xFF)
    default: (a, r, g, b) = (255, 0, 0, 0)
    }
    self.init(.sRGB, red: Double(r) / 255, green: Double(g) / 255, blue: Double(b) / 255, opacity: Double(a) / 255)
  }
}

enum CheckoutAppearance {
  static func makeDesignTokens(from options: NSDictionary) -> CheckoutComponents.DesignTokens? {
    let colorsDict = (options["colorTokens"] as? NSDictionary) ?? (options["colors"] as? NSDictionary)
    guard let colorsDict = colorsDict else { return nil }

    func hexColor(_ key: String) -> Color? {
      guard let hex = colorsDict[key] as? String, !hex.isEmpty else { return nil }
      return Color(hex: hex)
    }

    let action = hexColor("action") ?? Color(hex: "#186AFF")
    let background = hexColor("background") ?? Color(hex: "#FFFFFF")
    let border = hexColor("border") ?? Color(hex: "#DDDDDD")
    let disabled = hexColor("disabled") ?? Color(hex: "#B0B0B0")
    let error = hexColor("error") ?? Color(hex: "#AD283E")
    let formBackground = hexColor("formBackground") ?? Color(hex: "#FFFFFF")
    let formBorder = hexColor("formBorder") ?? Color(hex: "#949494")
    let inverse = hexColor("inverse") ?? Color(hex: "#FFFFFF")
    let outline = hexColor("outline") ?? Color(hex: "#91B2EE")
    let primary = hexColor("primary") ?? Color(hex: "#000000")
    let secondary = hexColor("secondary") ?? Color(hex: "#727272")
    let success = hexColor("success") ?? Color(hex: "#00856A")

    let colorTokens = CheckoutComponents.ColorTokens(
      action: action,
      background: background,
      border: border,
      disabled: disabled,
      error: error,
      formBackground: formBackground,
      formBorder: formBorder,
      inverse: inverse,
      outline: outline,
      primary: primary,
      secondary: secondary,
      success: success
    )

    let (radiusButton, radiusForm) = (parseBorderRadius(options["borderRadius"], defaultRadius: 4),
                                       parseBorderRadius(options["borderFormRadius"] ?? options["borderRadius"], defaultRadius: 4))

    let borderButtonRadius = CheckoutComponents.BorderRadius(radius: radiusButton, corners: .allCorners)
    let borderFormRadius = CheckoutComponents.BorderRadius(radius: radiusForm, corners: .allCorners)

    var designTokens = CheckoutComponents.DesignTokens(
      colorTokensMain: colorTokens,
      borderButtonRadius: borderButtonRadius,
      borderFormRadius: borderFormRadius
    )

    if let fontsDict = options["fonts"] as? NSDictionary {
      let fonts = parseFonts(fontsDict)
      if let fonts = fonts {
        designTokens = CheckoutComponents.DesignTokens(
          colorTokensMain: colorTokens,
          fonts: fonts,
          borderButtonRadius: borderButtonRadius,
          borderFormRadius: borderFormRadius
        )
      }
    }

    return designTokens
  }

  private static func parseBorderRadius(_ value: Any?, defaultRadius: CGFloat) -> CGFloat {
    guard let dict = value as? NSDictionary else { return defaultRadius }
    if let all = dict["all"] as? NSNumber {
      return CGFloat(all.doubleValue)
    }
    if let radius = dict["radius"] as? NSNumber {
      return CGFloat(radius.doubleValue)
    }
    return defaultRadius
  }

  private static func parseFonts(_ dict: NSDictionary) -> CheckoutComponents.DesignTokens.Fonts? {
    func makeFont(_ fontDict: NSDictionary?) -> CheckoutComponents.Font? {
      guard let d = fontDict else { return nil }
      let family = d["fontFamily"] as? String ?? "System"
      let size = (d["fontSize"] as? NSNumber)?.doubleValue ?? 17
      let lineHeight = (d["lineHeight"] as? NSNumber).map { CGFloat($0.doubleValue) }
      let letterSpacing = (d["letterSpacing"] as? NSNumber).map { CGFloat($0.doubleValue) }
      let font = Font.custom(family, size: size)
      return CheckoutComponents.Font(
        font: font,
        lineHeight: lineHeight ?? 4,
        letterSpacing: letterSpacing ?? 0
      )
    }
    let button = makeFont(dict["button"] as? NSDictionary)
    let footnote = makeFont(dict["footnote"] as? NSDictionary)
    let input = makeFont(dict["input"] as? NSDictionary)
    let label = makeFont(dict["label"] as? NSDictionary)
    let subheading = makeFont(dict["subheading"] as? NSDictionary)
    guard button != nil || footnote != nil || input != nil || label != nil || subheading != nil else { return nil }
    return CheckoutComponents.DesignTokens.Fonts(
      button: button ?? footnote ?? input ?? label ?? subheading!,
      footnote: footnote ?? button ?? input ?? label ?? subheading!,
      input: input ?? button ?? footnote ?? label ?? subheading!,
      label: label ?? button ?? footnote ?? input ?? subheading!,
      subheading: subheading ?? button ?? footnote ?? input ?? label!
    )
  }
}
